stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -394,6 +579,154 @@ public final class _ExecutionContext {
+      let address = buffer.advanced(by: idx)
+      // Each tensor can be symbolic (e.g. when using this API to create a
+      // symbolic input instance to tracee) or concrete (e.g. when creating the
+      // final output of the tracee).
+      let newHandle = TFE_TensorHandleCopySharingTensor(inputTensorHandle, status)
+      checkOk(status)
+      address.initialize(to: newHandle!)
+    }
+    TF_DeleteStatus(status)
+    self.init(_owning: buffer)
+  }
+}
+
+// TODO: assess if this protocol should be folded into TensorArrayProtocol.
+public protocol TensorArrayProtocolEnhanced : TensorArrayProtocol {
+  // Create an instance based on `inputs`, which can be symbolic (e.g. when
+  // creating a symbolic input to tracee) or concrete (e.g. when creating a
+  // final output of executing the tracee).
+  func createInstance(_owning inputs: [CTensorHandle]) -> Self
+}
+
+public func trace<State : TensorArrayProtocolEnhanced,
+                  Data : TensorGroup,
+                  Result : TensorGroup>(
+  with state: State,
+  in fn: (State, Data) -> (State, Result)
+) -> (State, Data) -> (State, Result) {
+  debugLog("""
+             Tracing over a function with \(state._tensorHandleCount) input \
+             state tensors and \(Data._typeList.count) input data tensors.
+             """)
+
+  // Verify that we are not already tracing.
+  internalConsistencyCheck(_RuntimeConfig.traceState.getFunction == nil,
+                           "Should not be in tracing mode already!")
+
+  // Switch to tracing mode.
+  let inputValueCount = Int(state._tensorHandleCount) + Data._typeList.count
+  _RuntimeConfig.traceState = .tracing(
+    TraceContext(inputValueCount: inputValueCount))
+
+  // Handle inputs.
+  let traceFn = _RuntimeConfig.traceState.getFunction!
+  // TODO: support other dtypes.
+  let inputSymbolicTensors = traceFn.symbolicInputs.map {
+    TFE_NewTensorHandleFromTFOutput($0, TF_FLOAT)!
+  }
+  internalConsistencyCheck(inputSymbolicTensors.count == inputValueCount)
+  let symbolicState = state.createInstance(
+    _owning: Array(inputSymbolicTensors.dropLast(Data._typeList.count)))
+  let symbolicData = Data(
+    _copying: Array(inputSymbolicTensors.dropFirst(
+                      Int(state._tensorHandleCount))))
+  // Run tracee to build the trace, adding ops to the trace graph function.
+  let (outputState, outputResult) = fn(symbolicState, symbolicData)
+
+  // The result is a closure that captures and executes the trace graph
+  // function.
+  return { (oldState: State, data: Data) -> (State, Result) in
+    debugLog("Running trace function over state \(oldState) and data \(data).")
+
+    // TAP means tensor array protocol.
+    let opType = "MyTraceFn_TAP"
+
+    debugLog("Getting input state tensor handles.")
+    let inputStateTensorHandles =  oldState._getCTensorHandles()
+    var inputTensors = inputStateTensorHandles.map {
+      Tensor<Float>(handle: TensorHandle(_owning: $0)) }
```

Comment:
Why is it specialized on `Float`?

Reply:
This is a short-cut we took for this initial PR. Left a `// TODO: support other dtypes.` above, and I repeated that TODO here.